### ISSUE TITLE
Use Meson subprojects to allow a system version of Boost, Qhull, & co

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#17751]
This can be done to make distro packagers happy, without introducing more issues around having extra build time dependencies or different versions of dependencies. https://mesonbuild.com/Subprojects.html#toggling-between-system-libraries-and-embedded-sources explains most of how that works. The one thing it does not show is how to restrict `dependency(...)` to a single version, but that is possible. That way, if the system has the right version of a library available already, it can be used. And if it's not found, we use our vendored copy (in the SciPy source tree or in a git submodule).

This came up several times:
- recently for Qhull, which conda-forge builds separately (see [this mailing list thread](https://mail.python.org/archives/list/scipy-dev@python.org/thread/I535LGRTWUXJIP2VM7VQGF3HELHQJ4B2/)),
- for Boost in gh-15260 (Boost is less interesting because it's header-only, but still some people seem to care)
- the Fortran-lang folks have asked about relying on their Minpack in the future (could be vendored still of course)
- there's other libraries that I can't remember questions about, but may be present in distros (ARPACK-ng, HiGHS, etc.)

This is not high-prio, but it would be nice to implement.

**Comment by rgommers** [comment]
xref https://github.com/mesonbuild/meson-python/discussions/410

**Comment by rgommers** [comment]
This blog post is a good example of using WrapDB for this kind of thing: https://nibblestew.blogspot.com/2023/05/advanced-dependency-management-and.html. It's even simpler than keeping a git submodule in `subprojects/`, but of course at the cost of needing network access if the dependency is not installed (seems fine in principle).

This could also be useful to start enforcing a result enough OpenBLAS version. Once there's a Meson OpenBLAS build and it's present in WrapDB (which is in the plan for next year), that can be the fallback when we specify `dependency('openblas', version='>=0.3.21')`. Right now we have no such check, and we do sometimes get bug reports from users on distros with several-years-old versions of OpenBLAS.

**Comment by doronbehar** [comment]
Are you planning to also support system wide installations of these dependencies?

**Comment by rgommers** [comment]
Yes, because distro packagers like that. The way it works now is that we vendor lots of stuff, and a small part of it gets unvendored in an ad-hoc way in some distros. The way we want it to work is to declare dependencies like so:
```
dependency('openblas', version: '>=0.3.21')
dependency('boost', modules : ['math'], version: '1.82')
```
and that would look for the system version first, and fall back on a vendored or downloaded version if no matching version was found on the system.

**Comment by rgommers** [comment]
Quick update: we should finally start shipping a subproject in 1.15.0, after gh-21565 HiGHS is one. We had to revert that once before, so I'd like to hold off on making more changes and see how it works out before doing the "look for system version first" dance.

**Comment by rgommers** [comment]
Boost is now searched for as a system dependency first, if the exact same version is found as we vendor in SciPy then the system version is used after the change in gh-21270.

**Comment by rgommers** [comment]
That should also align well with https://packaging.python.org/en/latest/discussions/downstream-packaging/#support-building-against-system-dependencies
